library(DESeq2)
library(data.table)
library(gtools)
library(rtracklayer)
library(GenomicFeatures)

where<-"3Prime-Seq/ReadCounts/"
setwd(where)

# Sample table for DEseq2 analysis
SampleInfo<-read.table("3Prime-Seq/SamplesInfo.txt", header=FALSE, stringsAsFactors=FALSE)
sampleFiles <- grep("ReadsPerGene.out.tab",list.files(where),value=TRUE)

## use the below to set the order of files
sampleFiles <- mixedsort(sampleFiles)

### for QuantSeq REV assay we choose 4th column of *ReadsPerGene.out.tab file generated by STAR equivalent to -s reverse is HTseq. 

countData = data.frame(fread(sampleFiles[1]))[c(1,4)]

# Loop and read the 4th column of remaining 11 *ReadsPerGene.out.tab files

for(i in 2:12) {
        countData = cbind(countData, data.frame(fread(sampleFiles[i]))[4])
}

sampleNames<-c("geneID",SampleInfo[,2])
colnames(countData) = sampleNames

# retrieving original head of STAR gene reads count data with sample stats for further summary

countDataHead<- countData[1:12,]   
rownames(countDataHead) = countDataHead[,1]
countDataHead<-countDataHead[,2:ncol(countDataHead)]

# Skip first 4 lines, count data starts on the 5th line
countData = countData[c(5:nrow(countData)),]
rownames(countData) = countData[,1]  ### moving geneIDs to rownames
countData = countData[,c(2:ncol(countData))] ### removing column with geneIDs

condition<-SampleInfo[,5]
replicate<-SampleInfo[,6]

# generating table with sample information
coldata = data.frame(cbind(condition, replicate))
coldata$type<-"single-read"
rownames(coldata) = sampleNames[2:length(sampleNames)]   # first name refers to geneIDs
coldata$condition <- factor(coldata$condition)
coldata$replicate <- factor(coldata$replicate)

# creating "DESeqDataSet" object for estimating sizing factors and further differential gene expression analysis
dds = DESeqDataSetFromMatrix(countData = countData, colData = coldata, design = ~ condition)

readCounts<-counts(dds, normalized = FALSE, replaced = FALSE)
readCountsDF<-as.data.frame(readCounts, stringsAsFactors = FALSE)
readCountsDF$id<-rownames(readCountsDF)

dds<-estimateSizeFactors(dds)
SF<-sizeFactors(dds)

readCountsNorm<-counts(dds, normalized = TRUE, replaced = FALSE)
readCountsNormDF<-as.data.frame(readCountsNorm, stringsAsFactors = FALSE)
readCountsNormDF$id<-rownames(readCountsNormDF)

#hg38 gtf file: /home/micgdu/GenomicData/genomesDec2022/gencode.v42.primary_assembly.annotation.gtf

# creating table with gene counts for all samples; adding gene info: coordinates, gene name, gene type
gtf<- import("/home/micgdu/GenomicData/genomesDec2022/gencode.v42.primary_assembly.annotation.gtf")
gtf_genes<-gtf[which(gtf$type=="gene")]
geneInfo<-as.data.frame(gtf_genes[,c(5,6,7)], stringsAsFactors = FALSE)
rownames(geneInfo)<-geneInfo$gene_id

# retrieving transcript length from gtf file 
gtf <- "/home/micgdu/GenomicData/genomesDec2022/gencode.v42.primary_assembly.annotation.gtf"
txdb.filename <- "v42.primary_assembly.hg38.txdb"
txdb <- makeTxDbFromGFF(gtf)
saveDb(txdb, txdb.filename)

hg38_txlens <- transcriptLengths(txdb)
head(hg38_txlens)


gene_transcript_length<- function(i,db) {
   median(db$tx_len[which(db$gene_id==gtf_genes$gene_id[i])])
}

Median_Transcript_length<-unlist(lapply(1:length(gtf_genes), gene_transcript_length, hg38_txlens))
geneInfo$txLens<-Median_Transcript_length

annotatedCounts<-cbind(geneInfo,readCountsDF)

save(annotatedCounts, file="3Prime-Seq/ReadCounts/annotatedCounts_3prime_ColoRectal.RData")

write.table(annotatedCounts, file="3Prime-Seq/ReadCounts/annotatedCounts_3prime_ColoRectal.txt",quote=FALSE,sep="\t",row.names=TRUE, col.names=TRUE)

# analogous table with gene counts normalized with sizing factors
annotatedNormCounts<-cbind(geneInfo,readCountsNormDF)
save(annotatedNormCounts, file="3Prime-Seq/ReadCounts/annotatedNormCounts_3prime_ColoRectal.RData")
write.table(annotatedNormCounts, file="3Prime-Seq/ReadCounts/annotatedNormCounts_3prime_ColoRectal.txt",quote=FALSE,sep="\t",row.names=TRUE, col.names=TRUE)

# analogous table with fpkm (fragment counts normalized per kilobase of feature length per million
# mapped fragments based on annotatedNormCounts and retrieved median transcript length
readCountsNormDF_FPKM<-readCountsNormDF
readCountsNormDF_FPKM$id<-NULL
readCountsNormDF_FPKM<-readCountsNormDF_FPKM/(geneInfo$txLens/1000)
annotatedNormCountsFPKM<-cbind(geneInfo,readCountsNormDF_FPKM)

save(annotatedNormCountsFPKM, file="3Prime-Seq/ReadCounts/annotatedNormCounts_FPKM_3prime_ColoRectal.RData")

write.table(annotatedNormCountsFPKM, file="3Prime-Seq/ReadCounts/annotatedNormCounts_FPKM_3prime_ColoRectal.txt",quote=FALSE,sep="\t",row.names=TRUE, col.names=TRUE)

# generating stats for the libraries:
sampleStats<-rbind(countDataHead[1:4,]/1e6,colSums(readCounts)/1e6)
featuredNorm<- sampleStats[5,]/SF
SF_rpm<-SF*mean(as.numeric(featuredNorm)) #sizing factors adjusted by single factor: mean level of rpm within exons in analyzed samples for convienience of track comparisons between experiment sets 
sampleStats<-rbind(colSums(sampleStats),sampleStats,featuredNorm, SF, SF_rpm)
rownames(sampleStats)<- c("all_reads", "N_unmapped","N_multimapping","N_noFeature","N_ambiguous","exons","exonsNorm", "SF", "SF_rpm")
sampleStats<-t(sampleStats)

write.table(sampleStats, file="3Prime-Seq/ReadCounts/Stats_3prime_ColoRectal.txt",quote=FALSE,sep="\t",row.names=TRUE, col.names=TRUE)

########### Normalising Stranded_bedgraphs
where<-"3Prime-Seq/Aligned/Stranded_bedGraphs/"
setwd(where)
sampleFiles <- grep(".bedgraph.gz",list.files(where),value=TRUE)
sampleFiles <- mixedsort(sampleFiles)

sampleNames<-rep(SampleInfo[,2], each = 2)
strand<- rep(a<-c("Fwd","Rev"), times=12)
SampleFactor<-rep(SF_rpm, each = 2)
sampleTable <- data.frame(fileName = sampleFiles, name=sampleNames, SampleFactor, strand)

bedtoolsBdgNorm<-function(x){
  # function based on sampleTable and DESeq2 sizing factors (SF) & stranded bedGraphs created before
  # all SF multiplied by the same factor: mean ~rpm of uniq reads in exons for analysed batch of samples
  # to faciliate vizual comparison between sample batches from different experiments   
  f<-sampleTable$SampleFactor[x]
  a<-read.table(sampleTable[x,1], header=FALSE, stringsAsFactors=FALSE)
  a[,2]<-format(a[,2], scientific=FALSE)
  a[,3]<-format(a[,3], scientific=FALSE)
  if (sampleTable$strand[x]=="Fwd") {
  a[,4]<-format(a[,4]/f,scientific=FALSE)} else { 
  a[,4]<-format(-a[,4]/f, scientific=FALSE)}
  a<-a[which(a[,1]%in% chr),]
  write.table(a ,file=paste(sampleTable$name[x], "_norm_",sampleTable$strand[x],".bedGraph", sep=""),quote=FALSE,sep="\t",row.names=FALSE, col.names=FALSE)
} 

chr<-c(paste("chr",1:22,sep=""),"chrX", "chrY", "chrM")      # limiting output to standard chromosomes

for (i in 1:dim(sampleTable)[1]) bedtoolsBdgNorm(i)
