library(DESeq2)
library(data.table)
library(gtools)
library(rtracklayer)
library(GenomicFeatures)

where<-"3Prime-Seq/ReadCounts/"
setwd(where)

# Sample table for DEseq2 analysis
SampleInfo<-read.table("3Prime-Seq/SamplesInfo.txt", header=FALSE, stringsAsFactors=FALSE)
sampleFiles <- grep("ReadsPerGene.out.tab",list.files(where),value=TRUE)

# Sort the order of files
sampleFiles <- mixedsort(sampleFiles)

# For QuantSeq REV assay we choose 4th column of *ReadsPerGene.out.tab file generated by STAR equivalent to -s reverse is HTseq

countData = data.frame(fread(sampleFiles[1]))[c(1,4)]

# Loop and read the 4th column of remaining 11 *ReadsPerGene.out.tab files

for(i in 2:12) {
        countData = cbind(countData, data.frame(fread(sampleFiles[i]))[4])
}

sampleNames<-c("geneID",SampleInfo[,2])
colnames(countData) = sampleNames

# Retrieve original head of STAR gene reads count data with sample stats for further summary
countDataHead<- countData[1:12,]   
rownames(countDataHead) = countDataHead[,1]
countDataHead<-countDataHead[,2:ncol(countDataHead)]

# Skip first 4 lines, as count data starts on the 5th line
countData = countData[c(5:nrow(countData)),]
rownames(countData) = countData[,1]  #geneIDs as rownames
countData = countData[,c(2:ncol(countData))] #remove the column with geneIDs

condition<-SampleInfo[,5]
replicate<-SampleInfo[,6]

# Sample Table creation
coldata = data.frame(cbind(condition, replicate))
coldata$type<-"single-read"
rownames(coldata) = sampleNames[2:length(sampleNames)]
coldata$condition <- factor(coldata$condition)
coldata$replicate <- factor(coldata$replicate)

# Create "DESeqDataSet" object to estimate size factors and normalisation
dds = DESeqDataSetFromMatrix(countData = countData, colData = coldata, design = ~ condition)
readCounts<-counts(dds, normalized = FALSE, replaced = FALSE)
dds2<-estimateSizeFactors(dds)
SF<-sizeFactors(dds2)

# Generate stats for the libraries:
sampleStats<-rbind(countDataHead[1:4,]/1e6,colSums(readCounts)/1e6)
featuredNorm<- sampleStats[5,]/SF
SF_rpm<-SF*mean(as.numeric(featuredNorm)) #Normalise the size factors based on the mean level of rpm within exons in the analyzed samples, makes dataasets more comparable
sampleStats<-rbind(colSums(sampleStats),sampleStats,featuredNorm, SF, SF_rpm)
rownames(sampleStats)<- c("all_reads", "N_unmapped","N_multimapping","N_noFeature","N_ambiguous","exons","exonsNorm", "SF", "SF_rpm")
sampleStats<-t(sampleStats)

write.table(sampleStats, file="3Prime-Seq/ReadCounts/Stats_3prime_ColoRectal.txt",quote=FALSE,sep="\t",row.names=TRUE, col.names=TRUE)

# Normalise Stranded_bedgraphs
where<-"3Prime-Seq/Aligned/Stranded_bedGraphs/"
setwd(where)
sampleFiles <- grep(".bedgraph.gz",list.files(where),value=TRUE)
sampleFiles <- mixedsort(sampleFiles)

sampleNames<-rep(SampleInfo[,2], each = 2)
strand<- rep(a<-c("Fwd","Rev"), times=12)
SampleFactor<-rep(SF_rpm, each = 2)
sampleTable <- data.frame(fileName = sampleFiles, name=sampleNames, SampleFactor, strand)

#bedtoolsBdgNorm - a funtion that normalises stranded bedGraphs using DESeq2 size factors (SF) and adjusts all SF by the mean rpm of exon reads for batch consistency to facilitate visual comparison between sample batches across experiments 

bedtoolsBdgNorm<-function(x){  
  f<-sampleTable$SampleFactor[x]
  a<-read.table(sampleTable[x,1], header=FALSE, stringsAsFactors=FALSE)
  a[,2]<-format(a[,2], scientific=FALSE)
  a[,3]<-format(a[,3], scientific=FALSE)
  if (sampleTable$strand[x]=="Fwd") {
  a[,4]<-format(a[,4]/f,scientific=FALSE)} else { 
  a[,4]<-format(-a[,4]/f, scientific=FALSE)}
  a<-a[which(a[,1]%in% chr),]
  write.table(a ,file=paste(sampleTable$name[x], "_norm_",sampleTable$strand[x],".bedGraph", sep=""),quote=FALSE,sep="\t",row.names=FALSE, col.names=FALSE)
} 

chr<-c(paste("chr",1:22,sep=""),"chrX", "chrY", "chrM") # limits the output to standard chromosomes

for (i in 1:dim(sampleTable)[1]) bedtoolsBdgNorm(i)
